flowzone.yml

```diff
@@ -1271,78 +1278,96 @@ jobs:
-          modified_files="$(git diff --name-only ; git ls-files --others --exclude-standard)"
-
-          # Extract changes
-          for file in $modified_files; do
-
-            echo "Creating blob of file $file..."
-            base64 -w0 "$file" > content.base64
-            response="$(gh api -X POST "/repos/$GH_REPO/git/blobs" \
-              -F 'content=@content.base64' \
-              -F 'encoding=base64')"
-            rm content.base64
-
-            echo "$response" | jq .
-            blob_sha="$(echo "$response" | jq -r .sha)"
-
-            # Add blob to our tree
-            tree_array+=("{\"path\":\"$file\",\"mode\":\"100644\",\"type\":\"blob\",\"sha\":\"$blob_sha\"}")
-          done
-
-          # Get the SHA of the tree the parent commit points to
-          base_tree_sha="$(git show -s --format=%T "$PARENT_COMMIT_SHA")"
-
-          # Create JSON array for tree creation
-          tree_json="$(printf ",%s" "${tree_array[@]}")"
-          tree_json="${tree_json:1}"
-          tree_json="[$tree_json]"
-          tree_json="{\"tree\": $tree_json, \"base_tree\": \"$base_tree_sha\"}"
-
-          echo "Creating tree..."
-          echo "$tree_json" | jq .
-
-          response="$(echo "$tree_json" | gh api -X POST "/repos/$GH_REPO/git/trees" --input -)"
+        uses: actions/github-script@60a0d83039c74a4aee543508d2ffcb1c3799cdea # v7
+        with:
+          github-token: ${{ steps.gh_app_token.outputs.token || secrets.FLOWZONE_TOKEN }}
+          script: |
+            const { execSync } = require('child_process');
+            const fs = require('fs');
+
+            // Get modified and untracked files
+            const getModifiedFiles = () => {
+              const modifiedCmd = 'git diff --name-only';
+              const untrackedCmd = 'git ls-files --others --exclude-standard';
+              const modified = execSync(modifiedCmd).toString().trim().split('\n');
+              const untracked = execSync(untrackedCmd).toString().trim().split('\n');
+              return [...new Set([...modified, ...untracked])].filter(f => f);
+            };
+
+            // Create tree entries for modified files
+            const createTreeEntries = async () => {
+              const entries = [];
+              const files = getModifiedFiles();
+
+              for (const file of files) {
+                core.info(`Creating blob for file ${file}...`);
+
+                // Read file content and create blob
+                const content = fs.readFileSync(file, "utf8");
+                const { data: blob } = await github.rest.git.createBlob({
+                  ...context.repo,
+                  content: Buffer.from(content).toString("base64"),
+                  encoding: 'base64'
+                });
+
+                entries.push({
+                  path: file,
+                  mode: '100644',
+                  type: 'blob',
+                  sha: blob.sha
+                });
+              }
```

Comment:
If the size of the files array is reasonable, you might be able to speed this up by using async `readFile` and `Promise.all`, something like

```
const entries = await Promise.all(files.map(async (file) => {
                core.info(`Creating blob for file ${file}...`);
                // Read file content and create blob
                const content = await fs.readFile(file, "utf8");
                const { data: blob } = await github.rest.git.createBlob({
                  ...context.repo,
                  content: Buffer.from(content).toString("base64"),
                  encoding: 'base64'
                });

                return {
                  path: file,
                  mode: '100644',
                  type: 'blob',
                  sha: blob.sha
                };
});
```

If the files array is too large, then you'll get rate limited, but you could use something like [what we do on base images](https://github.com/balena-io-library/base-images/blob/master/scripts/utils.js#L19-L36) to split the calls

Reply:
It should never be more than 4-6 files, I'll make the change.

Reply:
I pushed a new commit, if it looks correct and tests pass I'll squash it.